src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php

```diff
@@ -28,30 +28,70 @@
  */
 class ResponseCacheStrategy implements ResponseCacheStrategyInterface
 {
-    private $cacheable = true;
+    /**
+     * Cache-Control headers that are sent to the final response if they appear in ANY of the responses.
+     * @var array
+     */
+    private static $overrideDirectives = ['no-cache', 'no-store', 'no-transform', 'must-revalidate', 'proxy-revalidate'];
+
+    /**
+     * Cache-Control headers that are sent to the final response if the appear in ALL of the responses.
+     * @var array
+     */
+    private static $inheritDirectives = ['public', 'private', 'immutable'];
+
     private $embeddedResponses = 0;
-    private $ttls = array();
-    private $maxAges = array();
     private $isNotCacheableResponseEmbedded = false;
+    private $age = 0;
+    private $flagDirectives = array(
+        'no-cache' => null,
+        'no-store' => null,
+        'no-transform' => null,
+        'must-revalidate' => null,
+        'proxy-revalidate' => null,
+        'public' => null,
+        'private' => null,
+        'immutable' => null,
+    );
+    private $ageDirectives = array(
+        'max-age' => null,
+        's-maxage' => null,
+        'expires' => null,
+    );
 
     /**
      * {@inheritdoc}
      */
     public function add(Response $response)
     {
-        if (!$response->isFresh() || !$response->isCacheable()) {
-            $this->cacheable = false;
-        } else {
-            $maxAge = $response->getMaxAge();
-            $this->ttls[] = $response->getTtl();
-            $this->maxAges[] = $maxAge;
-
-            if (null === $maxAge) {
-                $this->isNotCacheableResponseEmbedded = true;
+        ++$this->embeddedResponses;
+
+        foreach (self::$overrideDirectives as $directive) {
+            if ($response->headers->hasCacheControlDirective($directive)) {
+                $this->flagDirectives[$directive] = true;
             }
         }
 
-        ++$this->embeddedResponses;
+        foreach (self::$inheritDirectives as $directive) {
+            if (false !== $this->flagDirectives[$directive]) {
+                $this->flagDirectives[$directive] = $response->headers->hasCacheControlDirective($directive);
+            }
+        }
+
+        $age = $response->getAge();
+        $this->age = max($this->age, $age);
+
+        if ($this->isUncacheable($response)) {
+            $this->isNotCacheableResponseEmbedded = true;
+            return;
+        }
+
+        $this->storeRelativeAgeDirective('max-age', $response->headers->getCacheControlDirective('max-age'), $age);
+        $this->storeRelativeAgeDirective('s-maxage', $response->headers->getCacheControlDirective('s-maxage'), $age);
+
+        $expires = $response->getExpires();
+        $expires = null === $expires ? null : $expires->format('U') - $response->getDate()->format('U');
```

Comment:
I suggest to swap the two expressions of the ternary op, thus checl `null !== $expire` to help reading with precedence